File: src/main/java/lee/a_合并两个有序链表.java
package lee;



import lee.model.ListNode;

import java.util.List;

public class a_合并两个有序链表 {

    /**
     * merge two lists
     * 优雅递归
     * @param list1 list1
     * @param list2 list2
     * @return {@link ListNode}
     */
    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if(list1 == null){
            return list2;
        }else if(list2 == null){
            return list1;
        }
        if(list1.val< list2.val){
            list1.next =mergeTwoLists(list1.next, list2);
            return list1;
        }else {
            list2.next =mergeTwoLists(list2.next, list1);
            return list2;
        }
    }

    /**
     * merge two lists1
     * 蔗民做法
     * @param list1 list1
     * @param list2 list2
     * @return {@link ListNode}
     */
    public static ListNode mergeTwoLists1(ListNode list1, ListNode list2) {
        ListNode help = null,head =null,save;
        if(list1 == null){
            return list2;
        }else if(list2 == null){
            return list1;
        }
        do{
            if(list1.val>list2.val){
                if(help!=null){
                    help.next = list2;
                }else {
                    head = list2;
                }
                save = list2.next;
                list2.next = list1;
                help = list2;
                list2 = save;
            } else {
                if(help!=null){
                    help.next =  list1;
                }else {
                    head = list1;
                }
                save = list1.next;
                list1.next = list2;
                help = list1;
                list1 = save;
            }
        }while (list1!=null && list2 !=null);
        return head;
    }
}

File: lee/src/main/java/org/example/lee/题目/链表/a_合并两个有序链表.go
package 链表

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	var a, head, c *ListNode
	if list1 == nil || (list2 != nil && list1.Val > list2.Val) {
		head = list2
	} else {
		head = list1
	}
	for list1 != nil && list2 != nil {
		//思路
		// c为要连接的点 如果 list1>list2
		// 那么c就要连接list2 ,然后c变为list2,list2变为next，继续循环
		if list1.Val > list2.Val {
			if c != nil {
				c.Next = list2
			}
			c = list2
			a = list2.Next
			list2.Next = list1
			list2 = a
		} else {
			if c != nil {
				c.Next = list1
			}
			c = list1
			a = list1.Next
			list1.Next = list2
			list1 = a
		}
	}
	return head
}
